File: tenumTalk07/27_removeElement_leetcode.js
/**
 * View challenge here:- https://leetcode.com/problems/remove-element/description/
 * Given an integer array nums and an integer val, remove all occurrences of val in nums in-place. The order of the elements may be changed. Then return the number of elements in nums which are not equal to val.

Consider the number of elements in nums which are not equal to val be k, to get accepted, you need to do the following things:

Change the array nums such that the first k elements of nums contain the elements which are not equal to val. The remaining elements of nums are not important as well as the size of nums.
Return k.
 * @param {number[]} nums
 * @param {number} val
 * @return {number}
 */
var removeElement = function (nums, val) {
  let i = 0;
  while (i != -1) {
    i = nums.indexOf(val);
    if (i != -1) {
      let removed = nums.pop();
      if (nums.length > 0 && i < nums.length) nums[i] = removed;
    }
  }
  return nums.length;
};

File: tenumTalk07/27_removeElement_leetcode.py
# https://leetcode.com/problems/remove-element/
"""
Assumptions:
* Ensure the array has at least one element.

Pseudocode:
- Keep track of two pointers (Variable i and j).
    - Their initial state is when (i) is at the start of the array while j is at the end.
- Iterate from the first element of the array till the last one.
    - Incase the current element is equal to the value to remove (Value k).
        - Swap the current element to where the element at index j is.
        - Reduce the space of j
        - Incement i only if the last element is not equal to k
    - Otherwise: Increment i
    NOTE: j should always be greater than or equal to i, otherwise, exit the iteration.
- i refers to the number of elements that are not equal to the value k
"""


def removeElement(nums: list[int], val: int) -> int:
    arrlen = len(nums)
    if arrlen == 0:
        return 0
    i = 0
    last_index = arrlen - 1
    while (i < arrlen):
        tmp = nums[i]
        if tmp == val:
            last_val = nums[last_index]
            nums[i] = last_val
            nums[last_index] = tmp
            if last_val != val:
                i += 1
            last_index -= 1
        else:
            i += 1
        # Incase the indices overlap
        if (last_index < i):
            break
    return i
